File: tests/sdk/go/s3_forensics_test.go
package sdk_test

import (
	"bytes"
	"context"
	"io"
	"strings"
	"testing"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	s3types "github.com/aws/aws-sdk-go-v2/service/s3/types"
)

func TestS3ForensicsObjectLockLite(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	client := newS3(t, cfg)
	ctx := context.Background()
	bucket := strings.ToLower(uniquePrefix(t) + "-lock")
	key := "locked.txt"

	_, err := client.CreateBucket(ctx, &s3.CreateBucketInput{
		Bucket:                     aws.String(bucket),
		ObjectLockEnabledForBucket: aws.Bool(true),
	})
	if err != nil {
		t.Fatalf("CreateBucket ObjectLock: %v", err)
	}
	t.Cleanup(func() {
		_, _ = client.DeleteObject(ctx, &s3.DeleteObjectInput{
			Bucket:                    aws.String(bucket),
			Key:                       aws.String(key),
			BypassGovernanceRetention: aws.Bool(true),
		})
		_, _ = client.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})

	retain := time.Now().UTC().Add(24 * time.Hour)
	_, err = client.PutObject(ctx, &s3.PutObjectInput{
		Bucket:                    aws.String(bucket),
		Key:                       aws.String(key),
		Body:                      bytes.NewReader([]byte("secret")),
		ObjectLockMode:            s3types.ObjectLockModeGovernance,
		ObjectLockRetainUntilDate: aws.Time(retain),
	})
	if err != nil {
		t.Fatalf("PutObject with Object Lock: %v", err)
	}

	_, err = client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	if err == nil {
		t.Fatal("DeleteObject without bypass expected AccessDenied")
	}

	_, err = client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket:                    aws.String(bucket),
		Key:                       aws.String(key),
		BypassGovernanceRetention: aws.Bool(true),
	})
	if err != nil {
		t.Fatalf("DeleteObject bypass governance: %v", err)
	}
}

func TestS3ForensicsBucketLogging(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	client := newS3(t, cfg)
	ctx := context.Background()
	prefix := uniquePrefix(t)
	src := strings.ToLower(prefix + "-src")
	logs := strings.ToLower(prefix + "-logs")
	key := "a.txt"

	for _, b := range []string{src, logs} {
		bucket := b
		_, err := client.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)})
		if err != nil {
			t.Fatalf("CreateBucket %s: %v", bucket, err)
		}
		t.Cleanup(func() {
			listed, _ := client.ListObjectsV2(ctx, &s3.ListObjectsV2Input{Bucket: aws.String(bucket)})
			if listed != nil {
				for _, obj := range listed.Contents {
					_, _ = client.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: obj.Key})
				}
			}
			_, _ = client.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
		})
	}

	_, err := client.PutBucketLogging(ctx, &s3.PutBucketLoggingInput{
		Bucket: aws.String(src),
		BucketLoggingStatus: &s3types.BucketLoggingStatus{
			LoggingEnabled: &s3types.LoggingEnabled{
				TargetBucket: aws.String(logs),
				TargetPrefix: aws.String("s3/"),
			},
		},
	})
	if err != nil {
		t.Fatalf("PutBucketLogging: %v", err)
	}

	gotLogging, err := client.GetBucketLogging(ctx, &s3.GetBucketLoggingInput{Bucket: aws.String(src)})
	if err != nil {
		t.Fatalf("GetBucketLogging: %v", err)
	}
	if gotLogging.LoggingEnabled == nil || aws.ToString(gotLogging.LoggingEnabled.TargetBucket) != logs {
		t.Fatalf("GetBucketLogging unexpected: %+v", gotLogging.LoggingEnabled)
	}

	_, err = client.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(src),
		Key:    aws.String(key),
		Body:   bytes.NewReader([]byte("abc")),
	})
	if err != nil {
		t.Fatalf("PutObject: %v", err)
	}

	listed, err := client.ListObjectsV2(ctx, &s3.ListObjectsV2Input{
		Bucket: aws.String(logs),
		Prefix: aws.String("s3/"),
	})
	if err != nil {
		t.Fatalf("ListObjectsV2 logs: %v", err)
	}
	if len(listed.Contents) < 1 {
		t.Fatalf("expected access log object under s3/, got %+v", listed.Contents)
	}
	obj, err := client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(logs),
		Key:    listed.Contents[0].Key,
	})
	if err != nil {
		t.Fatalf("GetObject log: %v", err)
	}
	raw, err := io.ReadAll(obj.Body)
	_ = obj.Body.Close()
	if err != nil {
		t.Fatalf("read log: %v", err)
	}
	line := string(raw)
	if !strings.Contains(line, "REST.PUT.OBJECT") || !strings.Contains(line, key) {
		t.Fatalf("access log line=%q", line)
	}
}

func TestS3ForensicsVersioningDeleteMarkers(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	client := newS3(t, cfg)
	ctx := context.Background()
	bucket := strings.ToLower(uniquePrefix(t) + "-ver")
	key := "obj.txt"

	_, err := client.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)})
	if err != nil {
		t.Fatalf("CreateBucket: %v", err)
	}
	t.Cleanup(func() {
		vers, _ := client.ListObjectVersions(ctx, &s3.ListObjectVersionsInput{Bucket: aws.String(bucket)})
		if vers != nil {
			for _, v := range vers.Versions {
				_, _ = client.DeleteObject(ctx, &s3.DeleteObjectInput{
					Bucket: aws.String(bucket), Key: v.Key, VersionId: v.VersionId,
				})
			}
			for _, m := range vers.DeleteMarkers {
				_, _ = client.DeleteObject(ctx, &s3.DeleteObjectInput{
					Bucket: aws.String(bucket), Key: m.Key, VersionId: m.VersionId,
				})
			}
		}
		_, _ = client.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})

	_, err = client.PutBucketVersioning(ctx, &s3.PutBucketVersioningInput{
		Bucket: aws.String(bucket),
		VersioningConfiguration: &s3types.VersioningConfiguration{
			Status: s3types.BucketVersioningStatusEnabled,
		},
	})
	if err != nil {
		t.Fatalf("PutBucketVersioning: %v", err)
	}

	put, err := client.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
		Body:   bytes.NewReader([]byte("payload")),
	})
	if err != nil {
		t.Fatalf("PutObject: %v", err)
	}
	if put.VersionId == nil || *put.VersionId == "" {
		t.Fatal("PutObject missing VersionId")
	}

	del, err := client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		t.Fatalf("DeleteObject: %v", err)
	}
	if del.DeleteMarker == nil || !*del.DeleteMarker {
		t.Fatalf("expected DeleteMarker=true, got %+v", del)
	}

	listed, err := client.ListObjectVersions(ctx, &s3.ListObjectVersionsInput{
		Bucket: aws.String(bucket),
		Prefix: aws.String(key),
	})
	if err != nil {
		t.Fatalf("ListObjectVersions: %v", err)
	}
	if len(listed.Versions) < 1 {
		t.Fatalf("expected object version, got %+v", listed.Versions)
	}
	if len(listed.DeleteMarkers) < 1 {
		t.Fatalf("expected delete marker, got %+v", listed.DeleteMarkers)
	}
}

File: tests/sdk/nodejs/test/s3-forensics.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CreateBucketCommand,
  DeleteBucketCommand,
  DeleteObjectCommand,
  GetBucketLoggingCommand,
  GetObjectCommand,
  ListObjectVersionsCommand,
  ListObjectsV2Command,
  PutBucketLoggingCommand,
  PutBucketVersioningCommand,
  PutObjectCommand,
} from "@aws-sdk/client-s3";
import { newS3, requireReady, uniquePrefix } from "../lib/helpers.mjs";

async function emptyAndDeleteBucket(s3, bucket) {
  try {
    const vers = await s3.send(
      new ListObjectVersionsCommand({ Bucket: bucket }),
    );
    for (const v of vers.Versions || []) {
      if (!v.Key) continue;
      try {
        await s3.send(
          new DeleteObjectCommand({
            Bucket: bucket,
            Key: v.Key,
            VersionId: v.VersionId,
            BypassGovernanceRetention: true,
          }),
        );
      } catch {
        /* ignore */
      }
    }
    for (const m of vers.DeleteMarkers || []) {
      if (!m.Key) continue;
      try {
        await s3.send(
          new DeleteObjectCommand({
            Bucket: bucket,
            Key: m.Key,
            VersionId: m.VersionId,
          }),
        );
      } catch {
        /* ignore */
      }
    }
  } catch {
    /* ignore */
  }
  try {
    const listed = await s3.send(new ListObjectsV2Command({ Bucket: bucket }));
    for (const obj of listed.Contents || []) {
      if (!obj.Key) continue;
      try {
        await s3.send(
          new DeleteObjectCommand({
            Bucket: bucket,
            Key: obj.Key,
            BypassGovernanceRetention: true,
          }),
        );
      } catch {
        /* ignore */
      }
    }
  } catch {
    /* ignore */
  }
  try {
    await s3.send(new DeleteBucketCommand({ Bucket: bucket }));
  } catch {
    /* ignore */
  }
}

test("S3 forensics Object Lock lite", async (t) => {
  if (!(await requireReady(t))) return;
  const s3 = newS3();
  const bucket = `${uniquePrefix()}-lock`.toLowerCase();
  const key = "locked.txt";

  await s3.send(
    new CreateBucketCommand({
      Bucket: bucket,
      ObjectLockEnabledForBucket: true,
    }),
  );
  t.after(() => emptyAndDeleteBucket(s3, bucket));

  const retain = new Date(Date.now() + 24 * 60 * 60 * 1000);
  await s3.send(
    new PutObjectCommand({
      Bucket: bucket,
      Key: key,
      Body: Buffer.from("secret"),
      ObjectLockMode: "GOVERNANCE",
      ObjectLockRetainUntilDate: retain,
    }),
  );

  await assert.rejects(
    () =>
      s3.send(
        new DeleteObjectCommand({
          Bucket: bucket,
          Key: key,
        }),
      ),
    /AccessDenied|Forbidden|denied/i,
  );

  await s3.send(
    new DeleteObjectCommand({
      Bucket: bucket,
      Key: key,
      BypassGovernanceRetention: true,
    }),
  );
});

test("S3 forensics bucket logging", async (t) => {
  if (!(await requireReady(t))) return;
  const s3 = newS3();
  const prefix = uniquePrefix();
  const src = `${prefix}-src`.toLowerCase();
  const logs = `${prefix}-logs`.toLowerCase();
  const key = "a.txt";

  await s3.send(new CreateBucketCommand({ Bucket: src }));
  await s3.send(new CreateBucketCommand({ Bucket: logs }));
  t.after(() => emptyAndDeleteBucket(s3, src));
  t.after(() => emptyAndDeleteBucket(s3, logs));

  await s3.send(
    new PutBucketLoggingCommand({
      Bucket: src,
      BucketLoggingStatus: {
        LoggingEnabled: {
          TargetBucket: logs,
          TargetPrefix: "s3/",
        },
      },
    }),
  );

  const got = await s3.send(new GetBucketLoggingCommand({ Bucket: src }));
  assert.equal(got.LoggingEnabled?.TargetBucket, logs);

  await s3.send(
    new PutObjectCommand({
      Bucket: src,
      Key: key,
      Body: Buffer.from("abc"),
    }),
  );

  const listed = await s3.send(
    new ListObjectsV2Command({ Bucket: logs, Prefix: "s3/" }),
  );
  assert.ok(
    (listed.Contents || []).length >= 1,
    `expected access log under s3/: ${JSON.stringify(listed.Contents)}`,
  );
  const obj = await s3.send(
    new GetObjectCommand({
      Bucket: logs,
      Key: listed.Contents[0].Key,
    }),
  );
  const line = Buffer.from(await obj.Body.transformToByteArray()).toString(
    "utf8",
  );
  assert.ok(line.includes("REST.PUT.OBJECT"), line);
  assert.ok(line.includes(key), line);
});

test("S3 forensics versioning delete markers", async (t) => {
  if (!(await requireReady(t))) return;
  const s3 = newS3();
  const bucket = `${uniquePrefix()}-ver`.toLowerCase();
  const key = "obj.txt";

  await s3.send(new CreateBucketCommand({ Bucket: bucket }));
  t.after(() => emptyAndDeleteBucket(s3, bucket));

  await s3.send(
    new PutBucketVersioningCommand({
      Bucket: bucket,
      VersioningConfiguration: { Status: "Enabled" },
    }),
  );

  const put = await s3.send(
    new PutObjectCommand({
      Bucket: bucket,
      Key: key,
      Body: Buffer.from("payload"),
    }),
  );
  assert.ok(put.VersionId, "PutObject missing VersionId");

  const del = await s3.send(
    new DeleteObjectCommand({ Bucket: bucket, Key: key }),
  );
  assert.equal(del.DeleteMarker, true);

  const listed = await s3.send(
    new ListObjectVersionsCommand({ Bucket: bucket, Prefix: key }),
  );
  assert.ok((listed.Versions || []).length >= 1, "missing object version");
  assert.ok(
    (listed.DeleteMarkers || []).length >= 1,
    "missing delete marker",
  );
});
